woocommerce/woocommerce
# Add order coupon dehydration

### Changes proposed in this Pull Request:

This PR does not completely address #36295. The PHP `serialize` function adds quite a bit of overhead in the resultant string with classes. For example a DateTime derived object is serialized to

```
O:11:"WC_DateTime":4:{s:13:"' . "\0*\0" . 'utc_offset";i:0;s:4:"date";s:26:"2023-01-30 18:27:51.000000";s:13:"timezone_type";i:1;s:8:"timezone";s:6:"+00:00";}
```

This PR converts the datetime fields to date strings and `WC_Meta_Data` objects to arrays. It also removes the non-boolean fields that default to falsey when they are empty. This reduces the footprint of a minimal coupon by ~60%.

> Observe thousands of rows of coupon_data occupying up to 94KB each.

Larger coupon rows reported in #36295 have a significant amount of data in their properties (eg. a large list of email addresses, added meta data, etc.). This PR will not affect the storage space for properties not outlined above. Addressing that will require a larger project to move the data to different storage.

See #36295



### How to test the changes in this Pull Request:

**Testing steps should be completed in the following scenarios**
1. In `trunk` & keep REST api output and screenshots for comparison with the remaining 3 scenarios
2. HPOS disabled
3. HPOS enabled with sync running
4. HPOS authoritative with sync disabled

**Prerequisites**
1. Create two coupons
  - Percentage
  - Fixed cart
  - Ensure one coupon has meta data. This can be added through the REST API
```
PUT wp-json/wc/v3/coupons/COUPONID
{"id":COUPONID,"meta_data":[{"key":"tag","value":"some-keyword"}]}
```
  - Ensure one coupon has an expiry date
2. Create an order with at least one product on it where both coupons will apply to the product

**Testing Steps**
1. Add the first coupon to the order & compare screen to trunk
2. Add a second product to the order & compare screen to trunk 
3. Click recalculate & compare to trunk
4. Refresh the screen & compare to trunk
5. Retrieve the order via the REST API & compare `coupon_lines` to trunk
6. When comparing the coupon properties in the REST API they do not need to be in the same order
7. Repeat the steps above with the second coupon

Test scenarios 2-4 should match the results found in `trunk`

### Other information:




### FOR PR REVIEWER ONLY:
